Add optional fetch size and direction to SqlQuery

nupa opened the pull request:
If the fetch size or fetch direction is set to
non-null value, it will be set to created
PreparedStatements.

resolves #36

nupa committed:
Add optional fetch size and direction to SqlQuery

If the fetch size or fetch direction is set to
non-null value, it will be set to created
PreparedStatements.

komu commented on dalesbred/src/main/java/org/dalesbred/Database.java:
```diff
@@ -743,6 +747,18 @@ private void logQueryExecution(@NotNull SqlQuery query, long millis) {
+        if (query.getFetchDirection() != null)
+            ps.setFetchDirection(query.getFetchDirection());
+        if (query.getFetchSize() != null)
+            ps.setFetchSize(query.getFetchSize());
```
I'm on a fence about these calls. Generally I dislike making assumptions that two successive calls return same value or that calls are cheap. Of course in this case we know that calling `getFetchDirection()` is cheap and that two successive calls will return the same value so the code is safe.

Still, I'd probably write these like:

```java
Integer fetchSize = query.getFetchSize();
if (fetchSize != null)
    ps.setFetchSize(fetchSize);
```

Sure, a couple of extra local variables, but less coupling (in the form of hidden assumptions) between the components.

That being said, for `fetchSize`, we could probably use `0` for "don't care" and make the type of the property `int`, since that is what JDBC uses as well. Now we have two different representations for "don't care": `null` and `0`. (Still I'd probably add `if (fetchSize != 0)` condition just to make sure we don't cause problems with funky JDBC drivers.)

komu commented on dalesbred/src/main/java/org/dalesbred/query/SqlQuery.java:
```diff
@@ -44,11 +46,17 @@
 
+    private @Nullable Integer fetchSize;
+
+    private @Nullable Integer fetchDirection;
```
Internally JDBC uses integer constants (because the API predates enums), but I'd add a new enumeration type for this:

```java
public enum FetchDirection {
    FORWARD(ResultSet.FETCH_FORWARD), 
    REVERSE(ResultSet.FETCH_REVERSE), 
    UNKNOWN(ResultSet.FETCH_UNKNOWN);

    private final int jdbcCode;

    FetchDirection(int jdbcCode) {
        this.jdbcCode = jdbcCode;
    }

    public int getJdcbCode() { 
        return jdbcCode;
    }
}
```

komu commented on dalesbred/src/main/java/org/dalesbred/query/SqlQuery.java:
```diff
@@ -112,6 +120,47 @@ private SqlQuery(@NotNull @SQL String sql, @NotNull List<?> args) {
+     */
+    public void setFetchSize(@Nullable Integer size) {
+        if (size != null && size < 0)
+            throw new IllegalArgumentException("Illegal fetch size " + size + ". Fetch size must be null or > 0");
```
The message should have _"or >= 0"_ instead of _"or > 0"_.

And looking at this, I think it's a good thing to get rid of `null` values here.

Before looking at the `PreparedStatement.setFetchSize` API I was about to comment that `setFetchSize(0)` makes no sense, but of course it does. It's probably a good idea to document the special meaning of zero here.

komu commented on dalesbred/src/main/java/org/dalesbred/query/SqlQuery.java:
```diff
@@ -112,6 +120,47 @@ private SqlQuery(@NotNull @SQL String sql, @NotNull List<?> args) {
+    }
+
+    /**
+     * Returns the fetch direction of ths query
```
Typo in "this". And let's add the trailing period while at it.

komu commented on dalesbred/src/main/java/org/dalesbred/query/SqlQuery.java:
```diff
@@ -112,6 +120,47 @@ private SqlQuery(@NotNull @SQL String sql, @NotNull List<?> args) {
+     */
+    public void setFetchDirection(@Nullable Integer direction) {
+        if (direction != null &&
+                !(ResultSet.FETCH_FORWARD == direction || ResultSet.FETCH_REVERSE == direction || ResultSet.FETCH_UNKNOWN == direction))
```
Having an enumeration will also simplify the implementation and documentation of this method by making illegal states statically impossible to represent.

komu commented on dalesbred/src/test/kotlin/org/dalesbred/SqlQueryTest.kt:
```diff
@@ -74,5 +78,21 @@ class SqlQueryTest {
+    }
+
+    @Test
+    fun illegalFetchDirection() {
```
This is ok, but you can also consider backticks, which allow you to rename the test methods more freely:

```kotlin
@Test
fun `illegal fetch direction`() {
    ...
}
```

I generally use this for new code, but most of Dalesbred's tests have been ported from the old Java code automatically so they retain their old naming.

nupa committed:
Add FetchDirection enum

ResultSet's fetch direction constants were ints,
now fetch direction values used in this API are always
valid.

komu commented on dalesbred/src/main/java/org/dalesbred/query/FetchDirection.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.sql.ResultSet;
+
+public enum FetchDirection {
```
Not that there's much to say about this, but I've tried to add some documentation for all user-facing program elements.

Perhaps something like:

> Represents fetch direction that can be given as a hint to {@link SqlQuery#setFetchDirection(FetchDirection)}.

And possiby link to `PreparedStatement.setFetchDirection` as well.

komu commented on dalesbred/src/main/java/org/dalesbred/query/SqlQuery.java:
```diff
@@ -44,11 +46,17 @@
         this.sql = requireNonNull(sql);
         this.args = unmodifiableList(args);
+        this.fetchSize = null;
+        this.fetchDirection = null;
```
These two rows are redundant, I'd leave them out.

nupa committed:
Add documentation

nupa committed:
Remove unnecessary initializations
